plugin/pkg/admission/nodedeclaredfeatures/admission.go

```diff
@@ -0,0 +1,183 @@
+	"k8s.io/component-base/featuregate"
+	ndf "k8s.io/component-helpers/nodedeclaredfeatures"
+	ndffeatures "k8s.io/component-helpers/nodedeclaredfeatures/features"
+	"k8s.io/kubernetes/pkg/apis/core"
+	v1 "k8s.io/kubernetes/pkg/apis/core/v1"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+const (
+	// PluginName is the name of this admission controller plugin.
+	PluginName = "NodeDeclaredFeatureValidator"
+)
+
+// Register registers a plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return NewPlugin()
+	})
+}
+
+// Plugin is an admission controller that validates pod updates against node capabilities.
+type Plugin struct {
+	*admission.Handler
+	nodeLister                     corev1listers.NodeLister
+	nodeDeclaredFeatureHelper      *ndf.Helper
+	nodeDeclaredFeatureGateEnabled bool
+}
+
+var _ admission.ValidationInterface = &Plugin{}
+var _ = genericadmissioninitializer.WantsExternalKubeInformerFactory(&Plugin{})
+var _ genericadmissioninitializer.WantsFeatures = &Plugin{}
+
+// New creates a new Plugin admission plugin
+func NewPlugin() (*Plugin, error) {
+	helper, err := ndf.NewHelper(ndffeatures.AllFeatures)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create node declared features helper: %w", err)
+	}
+	return &Plugin{
+		Handler:                   admission.NewHandler(admission.Create, admission.Update),
+		nodeDeclaredFeatureHelper: helper,
+	}, nil
+}
+
+// SetExternalKubeInformerFactory sets the informer factory for the plugin.
+func (p *Plugin) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	nodeInformer := f.Core().V1().Nodes()
+	p.nodeLister = f.Core().V1().Nodes().Lister()
+	p.SetReadyFunc(nodeInformer.Informer().HasSynced)
+}
+
+// SetFeatures sets the feature gates for the plugin.
+func (p *Plugin) InspectFeatureGates(featureGates featuregate.FeatureGate) {
+	p.nodeDeclaredFeatureGateEnabled = featureGates.Enabled(features.NodeDeclaredFeatures)
+}
+
+// ValidateInitialization ensures that the plugin is properly initialized.
+func (p *Plugin) ValidateInitialization() error {
+	if p.nodeLister == nil {
+		return fmt.Errorf("missing node lister for %s", PluginName)
+	}
+
+	if p.nodeDeclaredFeatureHelper == nil {
+		helper, err := ndf.NewHelper(ndffeatures.AllFeatures)
+		if err != nil {
+			return fmt.Errorf("failed to create node feature helper for %s: %w", PluginName, err)
+		}
+		p.nodeDeclaredFeatureHelper = helper
+	}
+	return nil
+}
+
+// Validate is the core of the admission controller logic.
+func (p *Plugin) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	// Ignore if the feature gate is not enabled.
+	if !p.nodeDeclaredFeatureGateEnabled {
+		return nil
+	}
+	// We only care about Pod updates.
+	if a.GetResource().GroupResource() != core.Resource("pods") {
+		return nil
+	}
+	if a.GetOperation() != admission.Update {
+		return nil
+	}
```

Comment:
This needs to filter by subresource. Some pod subresources don't have a pod as the request type (binding, exec, etc).